cmd/object-handlers.go

```diff
@@ -440,7 +443,7 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 		return
 	case authTypeAnonymous:
 		// http://docs.aws.amazon.com/AmazonS3/latest/dev/using-with-s3-actions.html
-		if s3Error := enforceBucketPolicy(bucket, "s3:PutObject", r.URL); s3Error != ErrNone {
+		if s3Error := enforceBucketPolicy(bucket, "s3:PutObject", *r); s3Error != ErrNone {
```

Comment:
OK, I was not very hopeful about protection against mutability here :) My attempt was only symbolic within the limits of the language.